daemon/cmd/device-reloader.go

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"slices"
+
+	"github.com/cilium/cilium/pkg/datapath/tables"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/hive/job"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+type deviceReloaderParams struct {
+	cell.In
+
+	Jobs          job.Registry
+	Scope         cell.Scope
+	DB            *statedb.DB
+	Daemon        promise.Promise[*Daemon]
+	Config        *option.DaemonConfig
+	Devices       statedb.Table[*tables.Device]
+	NodeAddresses statedb.Table[tables.NodeAddress]
+}
+
+type deviceReloader struct {
+	params       deviceReloaderParams
+	addrsChanged <-chan struct{}
+	devsChanged  <-chan struct{}
+	prevDevices  []string
+	jg           job.Group
+}
+
+// registerDeviceReloader provides the device reloader to the hive. The device reloader reloads the
+// datapath when the runtime devices (i.e. links) change. This is a legacy component. New code
+// should NOT be added here, but should directly use Table[*Device], Table[NodeAddress] or both.
+//
+// The current functionality of this reloader is to reload the datapath when either the node address
+// or the devices change. This leads to potentially inconsistent, observable state when the device
+// changes have not yet propagated to the NodeAddress changes. Components which depend on both need
+// to live with this fact and come up with component-specific strategies to deal with it.
+func registerDeviceReloader(lc hive.Lifecycle, p deviceReloaderParams) {
+	if !p.Config.EnableRuntimeDeviceDetection {
+		return
+	}
+
+	lc.Append(&deviceReloader{params: p})
+}
+
+// Start listening to changed devices if requested.
+func (d *deviceReloader) Start(ctx hive.HookContext) error {
+	if !d.params.Config.AreDevicesRequired() {
+		log.Info("Runtime device detection requested, but no feature requires it. Disabling detection.")
+		return nil
+	}
+
+	// Force an initial reload by supplying a closed channel.
+	c := make(chan struct{})
+	close(c)
+	d.devsChanged = c
+
+	jg := d.params.Jobs.NewGroup(d.params.Scope)
+	jg.Add(job.Timer("device-reloader", d.reload, time.Second))
```

Comment:
I see that the previous implementation used a for loop to trigger the reloading as soon as the devices were invalidated, while here we are running a periodic job once every `time.Second`. Why not use a OneShot job with a for loop to react immediately? This should also help in reducing the devices/addresses inconsistent window.
Also, from the job.Timer docstring:

`If the `fn` takes longer than the interval, its next invocation is not delayed.`

So, I'm wondering if it'd be possible for `relaod` to stay blocked on the select more than one second, leading to additional concurrent runs of `reload` itself.

Reply:
:) I know the `jobs.Timer` description isn't super clear about this, but it will _not_ run the same job concurrently. It'll just not delay the next execution.